pageserver/src/storage_sync.rs

```diff
@@ -1,103 +1,196 @@
+//! Yet instead of keeping the `metadata` file remotely, we wrap it with more
+//! data in an "index file" aka [`IndexPart`], containing the list of **all** remote
+//! files for a given timeline.
+//! If a file is not referenced from [`IndexPart`], it's not part of the remote storage state.
+//!
+//! Having the `IndexPart` also avoids expensive and slow `S3 list` commands.
+//!
+//! # Consistency
+//!
+//! To have a consistent remote structure, it's important that uploads and
+//! deletions are performed in the right order. For example, the index file
+//! contains a list of layer files, so it must not be uploaded until all the
+//! layer files that are in its list have been succesfully uploaded.
+//!
+//! The contract between client and its user is that the user is responsible of
+//! scheduling operations in an order that keeps the remote consistent as
+//! described above.
+//! From the user's perspective, the operations are executed sequentially.
+//! Internally, the client knows which operations can be performed in parallel,
+//! and which operations act like a "barrier" that require preceding operations
+//! to finish. The calling code just needs to call the schedule-functions in the
+//! correct order, and the client will parallelize the operations in a way that
+//! is safe.
+//!
+//! The caller should be careful with deletion, though. They should not delete
+//! local files that have been scheduled for upload but not yet finished uploading.
+//! Otherwise the upload will fail. To wait for an upload to finish, use
+//! the 'wait_completion' function (more on that later.)
+//!
+//! All of this relies on the following invariants:
+//!
+//! - We rely on read-after write consistency in the remote storage.
+//! - Layer files are immutable
+//!
+//! ## Implementation Note
+//!
+//! The *actual* remote state lags behind the *desired* remote state while
+//! there are in-flight operations.
+//! We keep track of the desired remote state in
+//! [`UploadQueueInitialized::latest_files`] and [`UploadQueueInitialized::latest_metadata`].
+//! It is initialized based on the [`IndexPart`] that was passed during init
+//! and updated with every `schedule_*` function call.
+//! All this is necessary necessary to compute the future [`IndexPart`]s
+//! when scheduling an operation while other operations that also affect the
+//! remote [`IndexPart`] are in flight.
+//!
+//! # Retries & Error Handling
+//!
+//! The client retries operations indefinitely, using exponential back-off.
+//! There is no way to force a retry, i.e., interrupt the back-off.
+//! This could be built easily.
+//!
+//! # Cancellation
+//!
+//! The operations execute as plain [`task_mgr`] tasks, scoped to
+//! the client's tenant and timeline.
+//! Dropping the client will drop queued operations but not executing operations.
+//! These will complete unless the `task_mgr` tasks are cancelled using `task_mgr`
+//! APIs, e.g., during pagserver shutdown or tenant detach.
```

Comment:
```suggestion
//! APIs, e.g., during pageserver shutdown or tenant detach.
```
This is a very nice comment by the way, exactly what I was hoping to find from these PRs :)

Reply:
Thanks :)